SSA/Projects/FullyHomomorphicEncryption/Syntax.lean

```diff
@@ -105,11 +91,39 @@ def ROfZComputable_impl (z : ℤ) : R q n :=
   }
   R.fromPoly p
 
+/-#
+Explanation of what's going on:
+- We want to have a `cstComputable : Z -> R`, which achieves three properties:
+1. It is computable, since the meta code needs it to be computable for `[fhe_icom|..]` to elaborate.
+2. It is equal to the noncomputable definition `(coe z : R q n)`
+3. It does not unfold when using `simp_peephole`, so we can rewrite using the equation
+     `ROfZComputable_def` after calling `simp_peephole`.
+
+What we really want is to have the following:
+
+```lean
 @[implemented_by ROfZComputable_impl]
-axiom ROfZComputable_stuck_term (q n : Nat) (z : ℤ) : R q n
+irreducible_noncomputable_def ROfZComputable_stuck_term (q n : Nat) (z : ℤ) :=
+  (↑ z : R q n)
+
+@[simp]
+lemma ROfZComputable_def (q n :Nat) (z : ℤ) : ROfZComputable_stuck_term  q n z = (↑ z : R q n) :=
+  ROfZComputable_stuck_term_def
+```
+
+However, the problem is that `irreducible_noncomputable_def` does not exist, so we fake it,
+by creating a computable `_implemented_by_shim`, and then making this shim an `irreducible_def`.
+-/
+
+@[implemented_by ROfZComputable_impl]
+def ROfZ_computable_stuck_term_implemented_by_shim (q n : Nat) (z : ℤ) := (↑ z : R q n)
+
+irreducible_def ROfZComputable_stuck_term (q n : Nat) (z : ℤ) :=
+  ROfZ_computable_stuck_term_implemented_by_shim q n z
```

Comment:
```suggestion
@[irreducible] def ROfZComputable_stuck_term (q n : Nat) (z : ℤ) :=
  ROfZ_computable_stuck_term_implemented_by_shim q n z
```
Unless I'm missing a reason why plain `@[irreducible] def` doesn't work, but it seems to do fine on machine when I tested locally. This way, we don't have to rely on the Mathlib version.

Reply:
Actually, I don't get it, how do I access the definition?